### ISSUE TITLE
Propshaft race condition with webpack creates 410 with dynamic chunk loading

### ISSUE THREAD

**Issue by joker-777** [rails/propshaft#110]
Hi, first of all, thanks a lot for developing this new solution. We are using it in production and are quite happy. One issue we bump into quite frequently though is in development.

We use webpack and depend on dynamic chunk loading. That means that webpack needs to compile these chunks already with a hash and a ".digested" suffix. We also use LiveReload from webpack to refresh the page after each compilation automatically.

What we observed is that when we apply multiple changes in succession, then a dynamic chunk request might return a 410. We debugged the code and could see that the asset wasn't in the [`@cached_assets_by_path`](https://github.com/rails/propshaft/blob/main/lib/propshaft/load_path.rb#L47). We also saw that propshaft clears the cache when any of the files in `app/assets/builds` changes. We suspect that there is some kind of race-condition between webpack and propshaft, clearing the cache, compiling multiple times and refreshing the browser at the same time.

We also noted that the digested assets created by webpack accumulate over time which could also slow down the creations of `cached_assets_by_path`. This accumulation happens also in production and doesn't get cleared by `assets:clear`.

We tried to find a reproducible example but since it is likely a race condition it is very hard. We would appreciate any thoughts and pointers.

**Comment by brenogazzola** [comment]
Hello @joker-777, thanks for trying Propshaft and raising the issue.

I'm not familiar with LiveReload, but from Propshaft side, it adds an extra `before_action` to the application controller of your app, that checks for changes and clears the cache when you make a new request. I think its possible that with a large enough number of assets, a second reload is going to hit while the cache of the first reload is being refreshed and this might cause the race condition you are seeing. I'll see what I can do.

As for production, `assets:clear` will keep at least three versions of each asset, and all that were created in the last hour. Unless you are deploying changes that touch all assets multiple times per hour, it should not be a problem. Also: production uses a manifest file, so reading files and calculating their digests dynamically.

**Comment by gingerlime** [comment]
Hi @brenogazzola! Thank you! I work with @joker-777 and we investigated this issue together.

We would really appreciate if you can look into it, and we're also happy to try out any ideas. As we mentioned, reproducing it is quite tricky. 

As for production, yes the manifest keeps track versions of assets and clears them on `assets:clear`. However assets that have a `.digested` in the filename aren't tracked or cleared by propshaft. We currently run a manual script to try to clear those up, but it's a bit brittle, because there's no easy way to know what were the last 3 versions, and we have to rely on created or updated timestamps of the files on the filesystem, but this is a crude tool (e.g. whatever the threshold we set for "old", if we don't deploy for longer than the threshold, we might risk pruning digested files that are still in-use).

**Comment by brenogazzola** [comment]
> we're also happy to try out any ideas

My two working ideas are using a semaphore when cache is reloading (which is going to have an impact on performance), or figuring out how to cache bust only the modified files (which makes code more complex).

> However assets that have a `.digested` in the filename aren't tracked or cleared by propshaft.

You mean from `builds` or `public/assets`? The `builds` folder needs to be commited, but it's contents should be in `.gitignore`. As for `public/assets`, that's definitely a bug. There was a recent change related to `.digested` and we must have missed it. I'll check.

> This is a crude tool

It is however how `assets:clean` work. It will be at least 3 versions of every asset, but as many as there were in the past 1 hour, all based in the `mtime`.

**Comment by gingerlime** [comment]
> My two working ideas are using a semaphore when cache is reloading (which is going to have an impact on performance), or figuring out how to cache bust only the modified files (which makes code more complex). Need to find time to test each of them.

I'm not sure about the more complex cache-busting, but I'd give semaphore a try, at least to see if it solves the issue we're seeing. I'll keep you posted if I find anything. That's a great idea! Thank you @brenogazzola.

About the `.digested` files, sorry. I explained the problem all wrong. Let me try to explain again :)

With "normal" assets, let's say `init.js`, propshaft automatically creates a digest and stores the file with the digest. So on production, you might end up with `init-aaaaa....js` and `init-bbbbb....js` etc. Then the cleanup process will map init.js to the two digested files. This allows propshaft to then clean up older versions from the filesystem. 

With ".digested" assets, propshaft gets an already-digested file, and doesn't apply another digest to it, so we might produce, e.g. `loader-abcdef.digested.js` and then during the next deploy `loader-fedcba.digested.js`. Propshaft does not extract the digest and logical part out of those files.

https://github.com/rails/propshaft/blob/390381548b125e8721c8aef1b9d894b7cc8bd868/lib/propshaft/output_path.rb#L55-L60

```
pry(main)> extract_path_and_digest("/app/loader-49b7fe393bf868bc53e952ba59a1120b037488edf92d60c70a9148ede3cdb326.digested.js")
=> ["/app/loader-49b7fe393bf868bc53e952ba59a1120b037488edf92d60c70a9148ede3cdb326.digested.js", nil]

pry(main)> extract_path_and_digest("/app/init-7324c15680c90f1782e2e2df9038c0c39b271a5d047745c84972e6f8635a4ddf.js")
=> ["/app/init.js", "7324c15680c90f1782e2e2df9038c0c39b271a5d047745c84972e6f8635a4ddf"]
```

This means that each `digested` file is a snowflake, and therefore never gets cleaned up, because there's only one `current` version of it. Our own crude tool is to just look for older `digested` files and delete them, but that's a foot-gun in many situations. I don't have a solution for this, but it's just something that we spotted as a limitation of the cleanup process when using digested assets.

Hope this makes sense? I'm not *that* familiar with propshaft yet, so I hope I didn't mess this up again :)

**Comment by gingerlime** [comment]
Yep! adding a mutex around the cache cleanup seems to solve the issue (fingers-crossed). I have very limited experience with concurrency in general, and basically zero with Ruby concurrency, but I simply wrapped the cleanup code in a mutex and it seems to do the trick

```ruby
class Propshaft::LoadPath
# ...

  def initialize(paths = [], version: nil)
    @paths   = dedup(paths)
    @version = version
    @mutex ||= Mutex.new
  end

# ...

    def clear_cache
      @mutex.synchronize do
        @cached_assets_by_path = nil
        assets_by_path  # seeds the cache
      end
    end

**Comment by tsrivishnu** [comment]
> This is a crude tool (e.g. whatever the threshold we set for "old", if we don't deploy for longer than the threshold, we might risk pruning digested files that are still in-use).

@gingerlime wouldn’t it be safer to parse the `.manifest.json` file and keep the files that are in there and delete all the rest?

**Comment by joker-777** [comment]
@tsrivishnu This is an option but then it would only keep one version. Shouldn't it keep at least the versions of the previous deployment?

**Comment by tsrivishnu** [comment]
I think it's worth noting that in the option 2, the manifest json that is versioned and kept is different from the `.manifest.json`.  I suppose those are generated by Webpack while `.manifest.json` is by propshaft. I haven't fully tested the options. But looks like option 2 can be used for pre-digested assets.

**Comment by dhh** [comment]
@gingerlime Could you try https://github.com/rails/propshaft/issues/110 and ensure that still works for you?

**Comment by dhh** [comment]
Re: production accumulation of digest chunks, is this an issue because you're using Heroku or something? With a containerized deployment, like Kamal, each container should be starting with a fresh build.

**Comment by gingerlime** [comment]
Thanks for following up on this. I’m no longer working on this codebase but @joker-777 should be able to comment on this issue.

**Comment by joker-777** [comment]
@dhh Thanks, I will try to test this as soon as possible.

**Comment by neontuna** [comment]
Wanted to leave a comment here and maybe help others with a similar issue that I think https://github.com/rails/propshaft/pull/189 has fixed.   Unfortunately currently if you search this error you get a lot of noise from people having basic issues with their <insert JS framework> tooling.

In development I would frequently get the following error in console and a white screen `such and such asset was blocked because of a disallowed MIME type ("text/plain")` and needed to hard refresh the browser (sometimes multiple times) in order to get the page to actually load.  This was essentially driving me insane as I worked on the front end because every other change I made would result in this issue.

<img width="1405" alt="Screenshot 2024-05-22 at 1 33 46 PM" src="https://github.com/rails/propshaft/assets/7414827/7417ee31-d25a-4c9a-910f-07a5f4143c5f">

There was never any errors from Webpack or the Rails console to go off - everything appeared fine otherwise.

I suspected there was some sort of race condition going on between propshaft, webpack, etc and had been following this thread.  We also kind of have a weird front end setup with Vue injected into erb templates, and tailwind running via Yarn for good measure.

So anyway, thank you @dhh - you saved me a lot of frustration :)

**Comment by joker-777** [comment]
@dhh The link in your comment doesn't work, but I guess you meant version 0.9.0. We haven't seen this error for a long time. It may be because we don't use LiveReload anymore. But we will start using 0.9.0 from now on and let you know if we find any problems.
Did you also find a solution for the `*.digested.js` files?

**Comment by dhh** [comment]
I did mean 0.9, yes. If you're talking about cleanup of digested files, I don't know that there is a straight shot there. It's also not a problem that affects modern containerized deployments, since they'll have a clean build every time.
